src/psyclone/gocean1p0.py

```diff
@@ -1865,3 +1877,118 @@ def data_on_device(self, parent):
+        :return: OpenCL language code representing the node.
+        :rtype: string
+        :raises GenerationError: if the kernel can not be converted to OpenCL
+        :raises NotImplementedError: if there are some symbol types or nodes
+                                     which are not implemented yet.
+        '''
+
+        # Assumptions made to generate GOcean OpenCL kernels.
+        # - It is a 2D domain.
+        # - First arguments are the iteration indices in column-major order.
+        # - All array have the same size and it is given by the
+        #   global_work_size argument to clEnqueueNDRangeKernel.
+        # - Assumes no dependencies among kernels called concurrently.
+
+        # TODO: At the moment, the method caller is responsible to ensure
+        # these assumptions. KernelSchedule access to the kernel
+        # meta-arguments could be used to check them and also improve the
+        # generated code. (Issue #288)
+
+        # Error checking
+        if len(self.symbol_table.argument_list) < 2:
+            raise GenerationError(
+                "GOcean Kernel should always have at least two "
+                "arguments representing the iteration indices.")
+
+        if (self.symbol_table.argument_list[0].datatype != "integer"
+                or len(self.symbol_table.argument_list[0].shape) != 0):
+            raise GenerationError(
+                "GOcean kernel first argument should be a scalar integer.")
+
+        if (self.symbol_table.argument_list[1].datatype != "integer"
+                or len(self.symbol_table.argument_list[1].shape) != 0):
+            raise GenerationError(
+                "GOcean kernel second argument should be a scalar integer.")
+
+        # Declare OpenCL kernel
+        code = self.indent(indent)
+        code = code + "__kernel "
+        code = code + "void " + self._name + "(\n"
+
+        # Generate kernel arguments, first 2 iteration indices are implicit.
+        array_arguments = []
+        for symbol in self.symbol_table.argument_list[2:]:
+            code = code + self.indent(indent + 1)
+            if len(symbol.shape) > 0:
+                code = code + "__global "
+
+            if symbol.datatype == "real":
+                code = code + "double "
+            elif symbol.datatype == "integer":
+                code = code + "int "
+            elif symbol.datatype == "character":
+                code = code + "char "
+            else:
+                raise NotImplementedError(
+                    "Type '{0}' OpenCL declaration not supported."
+                    "".format(symbol.datatype))
+
+            if len(symbol.shape) > 0:
+                code = code + "* restrict "
+                array_arguments.append((symbol.name, len(symbol.shape)))
+
+            code = code + symbol.name + ",\n"
+
+        code = code[:-2]   # Remove last ",\n"
+        code = code + "\n" + self.indent(indent + 1) + "){\n"
+
+        # Declare local variables.
```

Comment:
True, I created a local_symbols property in SymbolTable now.

Reply:
I meant that the whole business of declaring local symbols could be given over to the SymbolTable, rather than still doing each symbol separately here (which is a step towards what I was thinking).

Reply:
This (to my mind) is ~consistent with having each node in the PSyIR responsible for generating its own OpenCL representation. Since the SymbolTable is the only part of the PSyIR that represents declarations, it makes sense that it deals with the code-generation for them.

Reply:
I implemented gen_* methods for the code-generation of multiple `SymbolTable` elements. In this case local variables are `SymbolTable.gen_c_local_variables()`